core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -445,11 +445,15 @@ private RelDataType getTightestCommonTypeOrThrow(
     // 2. VARCHAR has 65536 as default precision.
     // 3. Following MS-SQL: BINARY or BOOLEAN can be casted to VARCHAR.
     if (SqlTypeUtil.isAtomic(type1) && SqlTypeUtil.isCharacter(type2)) {
-      resultType = factory.createSqlType(SqlTypeName.VARCHAR);
+      resultType =
+          factory.createTypeWithNullability(factory.createSqlType(SqlTypeName.VARCHAR),
```

Comment:
these changes are not really related to union.
But this may be the right approach.